pkg/vacuum/vacuum.go

```diff
@@ -275,35 +294,57 @@ func (e *Engine) Run(ctx context.Context) {
 		}
 		numberVacuumConnections.Set(0)
 	}()
+	// first we look for chunks past vacuum_freeze_min_age that probably should be frozen
 	// we limit ourselves to batches of 1000 chunks
 	// since we already have the advisory lock, continue to vacuum batches as needed until none left
 	for {
-		chunks, err := e.listChunks(ctx, con)
+		chunks, err := e.listChunks(ctx, con, sqlListChunksToFreeze)
 		if err != nil {
-			log.Error("msg", "failed to list chunks for vacuuming", "error", err)
-			return
+			log.Error("msg", "failed to list compressed chunks needing to be frozen", "error", err)
+			break
 		}
 		tablesNeedingVacuum.Set(float64(len(chunks)))
 		if len(chunks) == 0 {
-			log.Info("msg", "zero compressed chunks need to be vacuumed")
-			return
+			log.Info("msg", "zero compressed chunks needing to be frozen")
+			break
 		}
-		log.Info("msg", "compressed chunks need to be vacuumed", "count", len(chunks))
+		log.Info("msg", "compressed chunks needing to be frozen", "count", len(chunks))
 		numWorkers, err := e.calcNumWorkers(ctx, con, chunks)
 		if err != nil {
 			log.Error("msg", "failed to calc num workers", "error", err)
-			return
+			break
 		}
 		runWorkers(ctx, numWorkers, chunks, e.worker)
 		log.Debug("msg", "vacuum workers finished. delaying before next iteration...")
 		// in some cases, have seen it take up to 10 seconds for the stats to be updated post vacuum
 		time.Sleep(delay)
 	}
+	// next, we look for compressed chunks that have never been vacuum and seem to be missing
+	// statistics. autovacuum will ignore these until they hit vacuum_freeze_min_age, so let's
+	// catch these early
+	for {
```

Comment:
This for, I believe can starve out the loop above and vice-versa,.This should probably be one for loop that alternates between the two queries. Or maybe the loop above is more important and should starve-out this loop?

Reply:
take another look? @cevian